// react-redux/lib/components/createConnect.js
'use strict';

exports.__esModule = true;

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports['default'] = createConnect;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _utilsCreateStoreShape = require('../utils/createStoreShape');

var _utilsCreateStoreShape2 = _interopRequireDefault(_utilsCreateStoreShape);

var _utilsShallowEqualScalar = require('../utils/shallowEqualScalar');

var _utilsShallowEqualScalar2 = _interopRequireDefault(_utilsShallowEqualScalar);

var _utilsShallowEqual = require('../utils/shallowEqual');

var _utilsShallowEqual2 = _interopRequireDefault(_utilsShallowEqual);

var _utilsIsPlainObject = require('../utils/isPlainObject');

var _utilsIsPlainObject2 = _interopRequireDefault(_utilsIsPlainObject);

var _utilsWrapActionCreators = require('../utils/wrapActionCreators');

var _utilsWrapActionCreators2 = _interopRequireDefault(_utilsWrapActionCreators);

var _invariant = require('invariant');

var _invariant2 = _interopRequireDefault(_invariant);

var defaultMapStateToProps = function defaultMapStateToProps() {
  return {};
};
var defaultMapDispatchToProps = function defaultMapDispatchToProps(dispatch) {
  return { dispatch: dispatch };
};
var defaultMergeProps = function defaultMergeProps(stateProps, dispatchProps, parentProps) {
  return _extends({}, parentProps, stateProps, dispatchProps);
};

function getDisplayName(Component) {
  return Component.displayName || Component.name || 'Component';
}

function areStatePropsEqual(stateProps, nextStateProps) {
  var isRefEqual = stateProps === nextStateProps;
  if (isRefEqual || typeof stateProps !== 'object' || typeof nextStateProps !== 'object') {
    return isRefEqual;
  }

  return _utilsShallowEqual2['default'](stateProps, nextStateProps);
}

// Helps track hot reloading.
var nextVersion = 0;

function createConnect(React) {
  var Component = React.Component;
  var PropTypes = React.PropTypes;

  var storeShape = _utilsCreateStoreShape2['default'](PropTypes);

  return function connect(mapStateToProps, mapDispatchToProps, mergeProps) {
    var shouldSubscribe = Boolean(mapStateToProps);
    var finalMapStateToProps = mapStateToProps || defaultMapStateToProps;
    var finalMapDispatchToProps = _utilsIsPlainObject2['default'](mapDispatchToProps) ? _utilsWrapActionCreators2['default'](mapDispatchToProps) : mapDispatchToProps || defaultMapDispatchToProps;
    var finalMergeProps = mergeProps || defaultMergeProps;

    // Helps track hot reloading.
    var version = nextVersion++;

    return function (DecoratedComponent) {
      return (function (_Component) {
        _inherits(Connect, _Component);

        Connect.prototype.shouldComponentUpdate = function shouldComponentUpdate(nextProps, nextState) {
          return this.isSubscribed() && !areStatePropsEqual(this.state.stateProps, nextState.stateProps) || !_utilsShallowEqualScalar2['default'](this.props, nextProps);
        };

        _createClass(Connect, null, [{
          key: 'displayName',
          value: 'Connect(' + getDisplayName(DecoratedComponent) + ')',
          enumerable: true
        }, {
          key: 'DecoratedComponent',
          value: DecoratedComponent,
          enumerable: true
        }, {
          key: 'contextTypes',
          value: {
            store: storeShape.isRequired
          },
          enumerable: true
        }]);

        function Connect(props, context) {
          _classCallCheck(this, Connect);

          _Component.call(this, props, context);
          this.version = version;
          this.setUnderlyingRef = this.setUnderlyingRef.bind(this);
          this.state = _extends({}, this.mapState(props, context), this.mapDispatch(context));
        }

        Connect.prototype.isSubscribed = function isSubscribed() {
          return typeof this.unsubscribe === 'function';
        };

        Connect.prototype.trySubscribe = function trySubscribe() {
          if (shouldSubscribe && !this.unsubscribe) {
            this.unsubscribe = this.context.store.subscribe(this.handleChange.bind(this));
            this.handleChange();
          }
        };

        Connect.prototype.tryUnsubscribe = function tryUnsubscribe() {
          if (this.isSubscribed()) {
            this.unsubscribe();
            this.unsubscribe = null;
          }
        };

        Connect.prototype.componentDidMount = function componentDidMount() {
          this.trySubscribe();
        };

        Connect.prototype.componentWillUpdate = function componentWillUpdate() {
          if (process.env.NODE_ENV !== 'production') {
            if (this.version === version) {
              return;
            }

            // We are hot reloading!
            this.version = version;

            // Update the state and bindings.
            this.trySubscribe();
            this.setState(_extends({}, this.mapState(), this.mapDispatch()));
          }
        };

        Connect.prototype.componentWillUnmount = function componentWillUnmount() {
          this.tryUnsubscribe();
        };

        Connect.prototype.handleChange = function handleChange() {
          var props = arguments.length <= 0 || arguments[0] === undefined ? this.props : arguments[0];

          var nextState = this.mapState(props, this.context);
          if (!areStatePropsEqual(this.state.stateProps, nextState.stateProps)) {
            this.setState(nextState);
          }
        };

        Connect.prototype.mapState = function mapState() {
          var props = arguments.length <= 0 || arguments[0] === undefined ? this.props : arguments[0];
          var context = arguments.length <= 1 || arguments[1] === undefined ? this.context : arguments[1];

          var state = context.store.getState();
          var stateProps = finalMapStateToProps(state);

          _invariant2['default'](_utilsIsPlainObject2['default'](stateProps), '`mapStateToProps` must return an object. Instead received %s.', stateProps);

          return { stateProps: stateProps };
        };

        Connect.prototype.mapDispatch = function mapDispatch() {
          var context = arguments.length <= 0 || arguments[0] === undefined ? this.context : arguments[0];
          var dispatch = context.store.dispatch;

          var dispatchProps = finalMapDispatchToProps(dispatch);

          _invariant2['default'](_utilsIsPlainObject2['default'](dispatchProps), '`mapDispatchToProps` must return an object. Instead received %s.', dispatchProps);

          return { dispatchProps: dispatchProps };
        };

        Connect.prototype.merge = function merge() {
          var props = arguments.length <= 0 || arguments[0] === undefined ? this.props : arguments[0];
          var state = arguments.length <= 1 || arguments[1] === undefined ? this.state : arguments[1];
          var stateProps = state.stateProps;
          var dispatchProps = state.dispatchProps;

          var merged = finalMergeProps(stateProps, dispatchProps, props);

          _invariant2['default'](_utilsIsPlainObject2['default'](merged), '`mergeProps` must return an object. Instead received %s.', merged);

          return merged;
        };

        Connect.prototype.getUnderlyingRef = function getUnderlyingRef() {
          return this.underlyingRef;
        };

        Connect.prototype.setUnderlyingRef = function setUnderlyingRef(instance) {
          this.underlyingRef = instance;
        };

        Connect.prototype.render = function render() {
          return React.createElement(DecoratedComponent, _extends({ ref: this.setUnderlyingRef
          }, this.merge()));
        };

        return Connect;
      })(Component);
    };
  };
}

module.exports = exports['default'];
